File: 0062-unique-paths/0062-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[-1] * (n + 1) for _ in range(m + 1)]
        
        def recursion(row, column):
            if row == 1 or column == 1:
                return 1
            if dp[row][column] != -1:
                return dp[row][column]
            dp[row][column] = recursion(row - 1, column) + recursion(row, column - 1)
            return dp[row][column]
        
        return recursion(m, n)

File: 0062-unique-paths/0062-unique-paths.java
class Solution {
    public int uniquePaths(int m, int n) {
        return solve(0, 0, m, n, new int[m][n], 0);
    }

    public int solve(int x, int y, int m, int n, int[][] memo, int paths) {
        if (x == m-1 && y == n-1) return paths + 1;
        if (x > m-1 || y > n-1) return paths;
        if (memo[x][y] != 0) return memo[x][y];
        int i = solve(x+1, y, m, n, memo, paths);
        int j = solve(x, y+1, m, n, memo, paths);
        memo[x][y] = i+j;
        return i+j;
    }
}